### ISSUE TITLE
Custom Model Can Not Detection Object When Converted CoreML

### ISSUE THREAD

**Issue by akaomerr** [ultralytics/yolo-ios-app#30]
Hello, I trained my 8000-image dataset with YOLOv8m and it performed very well during testing. However, when I converted it to CoreML format, its performance decreased significantly and it misidentifies objects. Below, I'm sharing how I trained it and the code I used for converting to CoreML format.
`!yolo task=detect mode=train model=yolov5m.pt data={dataset.location}/data.yaml epochs=250 batch=32 imgsz=640 plots=True`
`from ultralytics import YOLO`
`model = YOLO("best.pt")`
`model.export(format="coreml",nms = True)`
<img width="1038" alt="Ekran Resmi 2024-06-07 15 15 56" src="https://github.com/ultralytics/yolo-ios-app/assets/87482709/83260568-6c31-474f-8a7a-6ac99c0b1f75">

**Comment by pderrenger** [comment]
Hello,

Thanks for reaching out and sharing the details of your issue. It seems like the performance drop might be related to the conversion process or model optimization for CoreML.

A few things to consider:
1. **Model Optimization**: Ensure that the model is fully optimized for CoreML. Sometimes, specific optimizations or layers might not translate perfectly from PyTorch to CoreML.
2. **NMS Settings**: You've enabled NMS in the export, which is good. However, tweaking the NMS thresholds might help in improving the detection accuracy.
3. **Model Verification**: After conversion, verify the model's layers and outputs to ensure they match the original PyTorch model. Discrepancies here can lead to performance issues.

If the problem persists, I recommend testing with a smaller subset of your data or a simpler model to isolate the issue. Also, reviewing the [CoreML model integration guide](https://docs.ultralytics.com/integrations/coreml/) might provide additional insights into best practices and troubleshooting tips.

**Comment by Logeshmuthu** [comment]
`model.export(format='mlmodel', nms=True)  `

convert using this parameter then load your model on xcode 
open your mode click utilities tab then click "update to model packge" button 
this method convert your model then run inference it works.

**Comment by akaomerr** [comment]
> hi @akaomerr `model.export(format='mlmodel', nms=True) `
> 
> convert using this parameter then load your model on xcode open your mode click utilities tab then click "update to model packge" button this method convert your model then run inference it works.

thanks for your advice, it's works 👍

**Comment by pderrenger** [comment]
Hello @akaomerr,

Thank you for sharing your solution! It's great to hear that updating the model to a package in Xcode resolved the issue. This is indeed a useful tip for ensuring the CoreML model performs as expected.

For anyone else encountering similar issues, it's always a good idea to:
1. **Verify Model Conversion**: Ensure the model is correctly converted by using the `model.export(format='mlmodel', nms=True)` command.
2. **Update in Xcode**: After loading the model in Xcode, use the "Update to Model Package" button in the Utilities tab to finalize the conversion.

If you continue to experience issues, please provide a minimum reproducible example so we can assist you better. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, always make sure you're using the latest version of the Ultralytics packages to benefit from the latest fixes and improvements.

Feel free to reach out if you have any more questions or need further assistance. Happy coding! 😊

**Comment by banqingyuan** [comment]
This question has been bugging me for three days!!! Ruined my weekend! I don't know why exporting directly to coreml causes the model to output incorrect results on iOS devices, but it works fine on the simulator, this is a problem that deserves to be taken seriously.

**Comment by pderrenger** [comment]
I understand your frustration and apologize for the inconvenience. This discrepancy might be due to hardware-specific behavior on iOS devices versus the simulator. Please ensure you are using the latest `ultralytics` version and update your CoreML model to a package in Xcode as described earlier. If the issue persists, testing with different CoreML configurations, such as adjusting `nms`, or verifying input preprocessing consistency between devices, may help. Let us know if further assistance is needed!

**Comment by shengyangyu** [comment]
> This question has been bugging me for three days!!! Ruined my weekend! I don't know why exporting directly to coreml causes the model to output incorrect results on iOS devices, but it works fine on the simulator, this is a problem that deserves to be taken seriously.

Hello, I encountered the same issue. Could you please tell me how you finally resolved it? Thank you.
